mcp/streamable.go

```diff
@@ -800,3 +857,83 @@ func (s *streamableClientConn) Close() error {
 	})
 	return s.closeErr
 }
+
+// establishSSE creates and manages the persistent SSE listening stream.
+// It is used for both the initial connection and all subsequent reconnection attempts,
+// using the Last-Event-ID header to resume a broken stream where it left off.
+// On a successful response, it delegates to handleSSE to process events;
+// on failure, it triggers the client's reconnection logic.
+func (s *streamableClientConn) establishSSE(opts *startSSEOptions) {
+	select {
+	case <-s.done:
+		return
+	default:
+	}
+
+	req, err := http.NewRequestWithContext(s.ctx, http.MethodGet, s.url, nil)
+	if err != nil {
+		return
+	}
+	s.mu.Lock()
+	if s._sessionID != "" {
+		req.Header.Set("Mcp-Session-Id", s._sessionID)
+	}
+	s.mu.Unlock()
+	if opts.lastEventID != "" {
+		req.Header.Set("Last-Event-ID", opts.lastEventID)
+	}
+	req.Header.Set("Accept", "text/event-stream")
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		// On connection error, schedule a retry.
+		s.scheduleReconnection(opts)
+		return
+	}
+
+	// Per the spec, a 405 response means the server doesn't support SSE streams at this endpoint.
+	if resp.StatusCode == http.StatusMethodNotAllowed {
+		resp.Body.Close()
+		return
+	}
+
+	if !strings.Contains(resp.Header.Get("Content-Type"), "text/event-stream") {
+		resp.Body.Close()
+		s.scheduleReconnection(opts)
+		return
+	}
+
+	s.handleSSE(resp, opts)
+}
+
+// scheduleReconnection schedules the next SSE reconnection attempt after a delay.
+func (s *streamableClientConn) scheduleReconnection(opts *startSSEOptions) {
+	reconnOpts := s.reconnectionOptions
+	if opts.attempt >= reconnOpts.MaxRetries {
+		return
```

Comment:
How does this manifest to the user? Is there a way to return an error?

Reply:
Good point, surfaced the error in s.err

Reply:
And updated Read to surface this error